linera-core/src/client/validator_manager/node_info.rs

```diff
@@ -0,0 +1,189 @@
+    ///
+    /// New nodes start with:
+    /// - 100ms expected latency (reasonable default)
+    /// - 100% success rate (optimistic start)
+    /// - Zero in-flight requests
+    /// - Default weights and smoothing factor
+    /// - Default normalization bounds
+    #[allow(unused)]
+    pub(super) fn new(node: RemoteNode<Env::ValidatorNode>) -> Self {
+        Self::with_config(
+            node,
+            ScoringWeights::default(),
+            0.1,
+            MAX_ACCEPTED_LATENCY_MS,
+            MAX_IN_FLIGHT_REQUESTS,
+        )
+    }
+
+    /// Creates a new `NodeInfo` with custom configuration.
+    pub(super) fn with_config(
+        node: RemoteNode<Env::ValidatorNode>,
+        weights: ScoringWeights,
+        alpha: f64,
+        max_expected_latency_ms: f64,
+        max_in_flight: usize,
+    ) -> Self {
+        Self {
+            node,
+            ema_latency_ms: 100.0, // Start with reasonable latency expectation
+            ema_success_rate: 1.0, // Start optimistically with 100% success
+            in_flight_requests: Arc::new(tokio::sync::RwLock::new(0)),
+            slot_available: Arc::new(tokio::sync::Notify::new()),
+            total_requests: 0,
+            weights,
+            alpha: alpha.clamp(0.01, 0.5), // Ensure alpha is in reasonable range
+            max_expected_latency_ms,
+            max_in_flight,
+        }
+    }
+
+    /// Calculates a normalized performance score (0.0 to 1.0) using weighted metrics.
+    ///
+    /// The score combines three normalized components:
+    /// - **Latency score**: Inversely proportional to EMA latency
+    /// - **Success score**: Directly proportional to EMA success rate
+    /// - **Load score**: Inversely proportional to current load
+    ///
+    /// Returns a score from 0.0 to 1.0, where higher values indicate better performance.
+    pub(super) async fn calculate_score(&self) -> f64 {
+        // 1. Normalize Latency (lower is better, so we invert)
+        let latency_score = 1.0
+            - (self.ema_latency_ms.min(self.max_expected_latency_ms)
+                / self.max_expected_latency_ms);
+
+        // 2. Success Rate is already normalized [0, 1]
+        let success_score = self.ema_success_rate;
+
+        // 3. Normalize Load (lower is better, so we invert)
+        let current_load = *self.in_flight_requests.read().await as f64;
+        let load_score =
+            1.0 - (current_load.min(self.max_in_flight as f64) / self.max_in_flight as f64);
+
+        // 4. Apply cold-start penalty for nodes with very few requests
+        let confidence_factor = (self.total_requests as f64 / 10.0).min(1.0);
+
+        // 5. Combine with weights
+        let raw_score = (self.weights.latency * latency_score)
+            + (self.weights.success * success_score)
+            + (self.weights.load * load_score);
+
+        // Apply confidence factor to penalize nodes with too few samples
```

Comment:
Is that necessary? If a node has too few samples, this makes us avoid getting more samples, doesn't it?